src/Bicep.Core/Parsing/Parser.cs

```diff
@@ -165,7 +165,7 @@ private SyntaxBase TargetScope(IEnumerable<SyntaxBase> leadingNodes)
         {
             var keyword = ExpectKeyword(LanguageConstants.TargetScopeKeyword);
             var assignment = this.WithRecovery(this.Assignment, RecoveryFlags.None, TokenType.NewLine);
-            var value = this.WithRecovery(() => this.Expression(allowComplexLiterals: true), RecoveryFlags.None, TokenType.NewLine);
+            var value = this.WithRecovery(() => this.Expression(allowComplexLiterals: true, allowResourceDeclarations: false), RecoveryFlags.None, TokenType.NewLine);

```

Comment:
I'm not super happy with how this shook out in the parser, and I'll need some help coming up with the right design. 

Now that objects are part of the expression parser, this change becomes pretty viral.

Reply:
You could create a `[Flags]` enum to replace the original variable. This way it's just one variable being passed around instead of two.

The other option would be to always allow the parser to parse resource declarations in all objects and then block it in the semantic checks. The downside of this would be that the resource keyword would be colorized in places where they're not really allowed.

---
In reply to: [577216127](https://github.com/Azure/bicep/pull/1516#discussion_r577216127) [](ancestors = 577216127)

Reply:
> The other option would be to always allow the parser to parse resource declarations in all objects and then block it in the semantic checks. The downside of this would be that the resource keyword would be colorized in places where they're not really allowed.

I'm not sure that would be an issue, since we're going to use lookahead for `:` to disambiguate with a property, a case like `{ resource: "foo" }` would still be parsed as a property.

Ultimately I didn't pursue this option because it just seemed like the wrong layer. We can represent the difference between "top level" and not using a grammar.